packages/venia-concept/src/components/ProductFullDetail/ProductFullDetail.js

```diff
@@ -135,11 +135,51 @@ class ProductFullDetail extends Component {
         );
     }
 
+    get mediaGalleryEntries() {
+        const { props, state } = this;
+        const { product } = props;
+        const { optionCodes, optionSelections } = state;
+        const {
+            configurable_options,
+            media_gallery_entries,
+            variants
+        } = product;
+        const isConfigurable = Array.isArray(configurable_options);
+
+        if (
+            !isConfigurable ||
+            (isConfigurable && optionSelections.size === 0)
+        ) {
+            return media_gallery_entries;
+        }
+
+        const item = variants.find(({ product: variant }) => {
+            for (const [id, value] of optionSelections) {
+                const code = optionCodes.get(id);
+
+                if (variant[code] !== value) {
+                    return false;
+                }
+            }
+
+            return true;
+        });
+
+        return item.product.media_gallery_entries;
+    }
+
     render() {
-        const { productOptions, props } = this;
+        const { productOptions, props, mediaGalleryEntries } = this;
         const { classes, product } = props;
         const { regularPrice } = product.price;
 
+        // We want this key to change whenever mediaGalleryEntries changes.
+        // Make it dependent on a unique value in each entry (file),
+        // and the order.
+        const carouselKey = mediaGalleryEntries.reduce((fullKey, entry) => {
```

Comment:
The media gallery entries object only changes when the actual product variant ID changes. The SKU can be the key here.

Reply:
I tried this but it didn't work because the SKU doesn't change between variants.

I was able to reproduce this bug if I only used the SKU:

![broken](https://user-images.githubusercontent.com/13182778/53360579-87ee5100-38fb-11e9-9952-68e1cb9535b5.gif)